Revise steps (#1715)
spectrocloud/librarium: docs/docs-content/clusters/public-cloud/aws/eks.md

Reviewer: The headings are a bit out of order. Prerequisites should be under Deploy an AWS Cluster. Or you can remove the heading Deploy an AWS Cluster and just use the main how-to headings.
![CleanShot 2023-11-02 at 09 44 00](https://github.com/spectrocloud/librarium/assets/29551334/a12253d0-9173-49b2-994f-7f253e9574fa)

Author: Fixed

18. Click on the **Validate** button and review the cluster configuration and settings summary. 

19. Click **Finish Configuration** to deploy the cluster. 

  The cluster details page of the cluster contains the status and details of the deployment. Use this page to track the deployment progress.
  
  :::info
  
  Provisioning an AWS EKS clusters can take several minutes.
  
  :::

You can access your Kubernetes cluster by using the kubectl CLI. Refer to the [Kubectl](../../cluster-management/palette-webctl.md) guide for more information.

### Validate

You can validate your cluster is up and in **Running** state.

1. Log in to [Palette](https://console.spectrocloud.com/).

2. Navigate to the left **Main Menu** and select **Clusters**. The **Clusters** page displays a list of all available clusters that Palette manages. 

3. Click on the cluster you created to view its details page.

4. Ensure the **Cluster Status** field displays **Running**.

<br />

## Enable Secrets Encryption for EKS Cluster

> Palette encourages using AWS Key Management Service (KMS) to provide envelope encryption of Kubernetes secrets stored in Amazon Elastic Kubernetes Service (EKS) clusters. This encryption is 

Author: Nice catch - yes.

a defense-in-depth security strategy to protect sensitive data such as passwords, docker registry credentials, and Transport Layer Security (TLS) keys stored as [Kubernetes Secrets](https://kubernetes.io/docs/concepts/configuration/secret/). 

> You can enable secrets encryption when you create an EKS cluster by toggling the **Enable encryption** button and providing the Amazon Resource Name (ARN) of the KMS key. The **Enable encryption** option is available on the cluster creation wizard's **Cluster Config** page for EKS.

Author: Added this sentence before the sentence in line 186: "Palette provides an **Enable encryption** option, which is only available during the cluster creation process."

### Prerequisites

- An AWS account added to Palette. Review [Add AWS Account](add-aws-accounts.md) for guidance.

- IAM user or role has attached policies listed in [Required IAM Policies](required-iam-policies.md).

> - A **PaletteControllersEKSPolicy** created in AWS and attached to the IAM user or role that Palette is using. To create this policy, refer to [Controllers EKS Policy](required-iam-policies.md#controllers-eks-policy).
> 
> 

Reviewer: ```suggestion
- A **PaletteControllersEKSPolicy** created in AWS and attached to the IAM user or role that Palette is using. To create this policy, refer to [Controllers EKS Policy](required-iam-policies.md#controllers-eks-policy).

- An AWS KMS key created in the AWS region you intended to deploy cluster to with Palette. 

- Ensure the KMS key policy allows the IAM user or role Palette usage of the KMS key. The KMS key policy must allowed the IAM role or IAM user the following actions:
      - "kms:Encrypt",
      - "kms:Decrypt",
      - "kms:ReEncrypt*",
      - "kms:GenerateDataKey*",
      - "kms:DescribeKey"
  
Check out the [Create a KMS Key policy](https://docs.aws.amazon.com/kms/latest/developerguide/key-policy-overview.html) guide for more information on key policies.     
          
```

Rita, I'm not sure how well my GitHub suggestion will go, so you might have to edit it locally 🙈

Author: Incorporated this block locally.

### Configure KMS Key

Use the following steps to configure a KMS key.

> 1. In AWS, locate the Key Management Service. 

Reviewer: ```suggestion
1. Log in to AWS console and navigateto the Key Management Service. 
```

> 2. Select the region where your Palette EKS cluster is deployed.

Reviewer: ```suggestion
2. Select the region where your KMS key policy is created.
```
This is all done before a cluster is created.

:::caution

> Ensure you create the KMS key in the same region as your Palette EKS cluster. Alternatively, you can create a multi-region KMS key that can be used across different regions. To learn how to create a multi-region key, review Amazon’s [Multi-Region Keys in AWS KMS](https://docs.aws.amazon.com/kms/latest/developerguide/multi-region-keys-overview.html) reference guide.

Reviewer: ```suggestion
Ensure you create the KMS key in the same region as you intend to deploy EKS clusters through Palette. Alternatively, you can create a multi-region KMS key that can be used across different regions. To learn how to create a multi-region key, review Amazon’s [Multi-Region Keys in AWS KMS](https://docs.aws.amazon.com/kms/latest/developerguide/multi-region-keys-overview.html) reference guide.

```

::: 

3. Create a key of type **Symmetric** and with usage **Encrypt and decrypt**.

> 4. Ensure the IAM user or role that Palette is using has a policy attached with the following required IAM permissions. 

Reviewer: ```suggestion
4. Ensure the IAM user or role that Palette is using has a policy attached with the following required IAM permissions.  Replace the account ID and `REPLACE_ME` with the name of IAM User. If you are using an IAM role, change the ARN to end with `:role/REPLACE_ME`.

```

> 4. Ensure the IAM user or role that Palette is using has a policy attached with the following required IAM permissions. 
> 
>   ```json hideClipboard
>   kms:CreateGrant,
>   kms:ListAliases,
>   kms:ListKeys,
>   kms:DescribeKeys
>   ```

Reviewer: ```suggestion
4. Ensure the IAM user or role that Palette is using has a policy attached with the following required IAM permissions. 
```
We can simply this by just adding the policy below.

   
   Example:

   ```json

>    {
>     "Version": "2012-10-17",
>     "Statement": [
>         {
>             "Sid": "Statement1",
>             "Effect": "Allow",
>             "Action": [
>                 "kms:ListKeys",
>                 "kms:ListAliases",
>                 "kms:DescribeKey",
>                 "kms:CreateGrant"
>             ],
>             "Resource": "*"
>         }
>     ]
>   }

Reviewer: ```suggestion
  {
      "Sid": "Allow Palette use of the KMS key",
      "Effect": "Allow",
      "Principal": {
          "AWS": "arn:aws:iam::123456789:user/REPLACE_ME"
      },
      "Action": [
          "kms:Encrypt",
          "kms:Decrypt",
          "kms:ReEncrypt*",
          "kms:GenerateDataKey*",
          "kms:DescribeKey"
      ],
      "Resource": "*"
  },
```
This what the key policy should look like. The user just needs replace the account ID `123456789` and the user name/role name at the end.

   ```

>   

Reviewer: ```suggestion

  :::info 
  
  If you are using IAM to delegate access to the KMS key, you can continue to do so without modifying the KMS key 
  policy. Ensure the Palette IAM User or role have the proper custom IAM policy attached that grants it access to the 
  KMS key. Refer to the [Using IAM policies with AWS KMS](https://docs.aws.amazon.com/kms/latest/developerguide/iam-policies.html) to learn more about managing KMS key with IAM policies. 
  :::  

```

This is an advanced topic but very common in large organizations that have good controls.

a defense-in-depth security strategy to protect sensitive data such as passwords, docker registry credentials, and Transport Layer Security (TLS) keys stored as [Kubernetes Secrets](https://kubernetes.io/docs/concepts/configuration/secret/). 

You can enable secrets encryption when you create an EKS cluster by toggling the **Enable encryption** button and providing the Amazon Resource Name (ARN) of the KMS key. The **Enable encryption** option is available on the cluster creation wizard's **Cluster Config** page for EKS.

### Prerequisites

- An AWS account added to Palette. Review [Add AWS Account](add-aws-accounts.md) for guidance.

- IAM user or role has attached policies listed in [Required IAM Policies](required-iam-policies.md).

- A **PaletteControllersEKSPolicy** created in AWS and attached to the IAM user or role that Palette is using. To create this policy, refer to [Controllers EKS Policy](required-iam-policies.md#controllers-eks-policy).

### Configure KMS Key

Use the following steps to configure a KMS key.

> 1. Log in to AWS console and navigateto the Key Management Service. 

Author: fixed

  |**Private Access CIDRs** |This setting controls which private IP address CIDR ranges can access the cluster. Private CIDRs provide a way to specify private, self-hosted, and air-gapped networks or Private Cloud Gateway (PCG) that may be located in other VPCs connected to the VPC hosting the cluster endpoint.<br /><br />To restrict network access, enter the IP address CIDR range that will provide access to the cluster. Although `0.0.0.0/0` is pre-populated in this field, only IPs that can reach the private endpoint are those within the VPC or any other connected VPCs. For example, while using `0.0.0.0/0` would allow traffic throughout the VPC and all peered VPCs, specifying the VPC CIDR `10.0.0.0/16` would limit traffic to an individual VPC. For more information, refer to the [Amazon EKS cluster endpoint access control](https://docs.aws.amazon.com/eks/latest/userguide/cluster-endpoint.html) reference guide.|

>   |**Enable Encryption**| Use this option for secrets encryption. You must have an existing AWS Key Managment Service (KMS) key you can use. Toggle the **Enable encryption** option and use the **drop-down Menu** in the **ARN** field to select the KMS key ARN.<br /><br />If you do not have a KMS key and want to create one to use this option, review [Enable Secrets Encryption for EKS Cluster](enable-secrets-encryption-kms-key.md). Once your KMS key is created, return to this Cluster Config step to enable secrets encryption and specify the KMS key ARN. |

Author: fixed typo.

---
sidebar_label: "Create and Manage AWS EKS Cluster"
title: "Create and Manage AWS EKS Cluster"
description: "Learn how to deploy and manage AWS EKS clusters with Palette."
hide_table_of_contents: false
tags: ["public cloud", "aws", "eks"]
sidebar_position: 30
---

Palette supports creating and managing Amazon Web Services (AWS) Elastic Kubernetes Service (EKS) clusters deployed to an AWS account. This section guides you on how to create an EKS cluster in AWS that Palette manages.

## Prerequisites

- Access to an AWS cloud account.

- Palette integration with AWS account. Review [Add AWS Account](add-aws-accounts.md) for guidance.

- An infrastructure cluster profile for AWS EKS. When you create the profile, ensure you choose **EKS** as the **Managed Kubernetes** cloud type. Review [Create an Infrastructure Profile](../../../profiles/cluster-profiles/create-cluster-profiles/create-infrastructure-profile.md) for guidance.

- An EC2 key pair for the target region that provides a secure connection to your EC2 instances. To learn how to create a key pair, refer to the [Amazon EC2 key pairs](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-key-pairs.html) resource.

> - kubelogin installed. This is a kubectl plugin for Kubernetes OpenID Connect (OIDC) authentication, also known as `kubectl oidc-login`.

Author: added link.

> - Palette creates compute, network, and storage resources in AWS when it provisions Kubernetes clusters. Ensure there is sufficient capacity in the preferred AWS region to create the following resources:

Reviewer: This is true, but the user has the ability to specify an existing VPC. If they provide their own VPC, then Palette won't create the resources below.

Author: added clarification.

- An infrastructure cluster profile for AWS EKS. When you create the profile, ensure you choose **EKS** as the **Managed Kubernetes** cloud type. Review [Create an Infrastructure Profile](../../../profiles/cluster-profiles/create-cluster-profiles/create-infrastructure-profile.md) for guidance.

- An EC2 key pair for the target region that provides a secure connection to your EC2 instances. To learn how to create a key pair, refer to the [Amazon EC2 key pairs](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-key-pairs.html) resource.

- kubelogin installed. This is a kubectl plugin for Kubernetes OpenID Connect (OIDC) authentication, also known as `kubectl oidc-login`.

- Palette creates compute, network, and storage resources in AWS when it provisions Kubernetes clusters. Ensure there is sufficient capacity in the preferred AWS region to create the following resources:
    - Virtual CPU (vCPU)
    - Virtual Private Cloud (VPC)
    - Elastic IP
    - Internet Gateway
    - Elastic Load Balancers
    - Network Address Translation (NAT) Gateway

:::info

> To enable automatic subnet discovery for integration with AWS load balancer service, you need to add tags to the the Virtual Private Cloud (VPC) public subnets. Use the AWS Tag Editor and specify the region and resource type. Then, add the following tags. Replace the value `yourClusterName` with your cluster's name. Refer to [AWS Tag Editor](https://docs.aws.amazon.com/tag-editor/latest/userguide/tag-editor.html) for more information.

Author: fixed.

> To enable automatic subnet discovery for integration with AWS load balancer service, you need to add tags to the the Virtual Private Cloud (VPC) public subnets. Use the AWS Tag Editor and specify the region and resource type. Then, add the following tags. Replace the value `yourClusterName` with your cluster's name. Refer to [AWS Tag Editor](https://docs.aws.amazon.com/tag-editor/latest/userguide/tag-editor.html) for more information.

Reviewer 2: "with **the** AWS load balancer service" 
---
"To enable automatic subnet discovery for integration with AWS load balancer service, you need to add tags to the the Virtual Private Cloud (VPC) public subnets."
what about:
"To enable the automated creation of AWS external load balancers, it is required to tag all public subnets in the Virtual Private Cloud (VPC) public subnets."

Author: I'm not sure about this suggestion. It sounds like the subnets need to be discovered and tags are the means to locate them. I'll check with @karl-cardenas-coding.

> <!-- The following tags should be added to the virtual private network (VPC) public subnets to enable automatic subnet discovery for integration with AWS load balancer service. Replace the value `yourClusterName` with your cluster's name. -->

Reviewer 2: I found this article while reading this PR:
https://repost.aws/knowledge-center/eks-vpc-subnet-discovery

Author: Added a link to it - thank you.

13. Click **Finish Configuration** to deploy the cluster. 

  The cluster details page of the cluster contains the status and details of the deployment. Use this page to track the deployment progress.
  
    #### Effect Table
    
    |**Parameter**| **Description**|
    |-------------|---------------|
    | **NoSchedule**|  A pod that cannot tolerate the node taint and should not be scheduled to the node. 
    | **PreferNoSchedule**| The system will avoid placing a non-tolerant pod to the tainted node but is not guaranteed.
    | **NoExecute**|  New pods will not be scheduled on the node, and existing pods on the node if any on the node will be evicted they do not tolerate the taint. |
  :::info
  
  Provisioning an AWS EKS clusters can take several minutes.
  
  :::

14. Click on **Next**.  
    
15. The settings page is where you can configure patching schedule, security scans, backup settings, setup role based access control (RBAC), and enable [Palette Virtual Clusters](../../../devx/palette-virtual-clusters/palette-virtual-clusters.md). Review the settings and make changes if needed. Click on **Validate**.

> 16. Review the settings summary and click on **Finish Configuration** to deploy the cluster. Be aware that provisioning an AWS EKS clusters can take several minutes.

Reviewer: A question we see a lot is how to connect to the cluster. We could add a sentence or two about this and link to the kubectl page we have is cluster management.

Author: Added sentence with a link.

## Deploy an AWS Cluster

> Use the following steps to deploy an AWS cluster in which to provision an EKS cluster.

Reviewer 2: I don't think this is right. The following steps will deploy an EKS cluster on AWS (or similar).

Author: Fixed.

3. From the left **Main Menu** select **Clusters**, and click on the **Add New Cluster** button.

4. Select **Deploy New Cluster** on the next page Palette displays. This will allow you to deploy a cluster using your own cloud account. 

5. Select **AWS** and click on the **Start AWS Configuration** button.

6. Fill out the following basic information, and click **Next** to continue.

  | **Field** | **Description** |
  |-----------|-----------------|
  | **Cluster Name**| A custom name for the cluster. |
  | **Description**| Use the description to provide context about the cluster.|
  | **Tags**| Assign any desired cluster tags. Tags on a cluster are propagated to the Virtual Machines (VMs) deployed to the target environments. Example: `region:us-east-1a` or `zone:vpc-private-us-east-1a`.|
  | **Cloud Account** | If you already added your AWS account in Palette, select it from the **drop-down Menu**. Otherwise, click on **Add New Account** and add your AWS account information. |

  To learn how to add an AWS account, review the [Add an AWS Account to Palette](add-aws-accounts.md) guide.

### Create an EKS Cluster

Use the following steps to provision a new EKS cluster.

> 1. Select **EKS** as the **Managed Kubernetes**. 

Reviewer 2: this sounds clunky?

Author: I agree - I debated how to guide the user in the UI. Changed it to "Select **EKS** listed under **Managed Kubernetes**.

2. Select the EKS cluster profile you created and click on **Next**. Palette displays the cluster profile layers.

3. Review the profile layers and customize parameters as desired in the YAML files that display when you select a layer.

>   You can configure OpenID Connect (OIDC) at the Kubernetes layer. To configure OIDC for managed EKS clusters, follow steps for Amazon EKS in the [Configure Custom OIDC](../../../integrations/kubernetes.md/#configure-custom-oidc) guide.

Author: fixed.

  :::caution

  Configuring OIDC requires you to map a set of users or groups to a Kubernetes RBAC role. To learn how to map a Kubernetes role to users and groups, refer to [Create Role Bindings](../../cluster-management/cluster-rbac.md/#create-role-bindings). Refer to [Use RBAC with OIDC](../../../integrations/kubernetes.md/#use-rbac-with-oidc) for an example.

  :::

4. Click on **Next** to continue.

5. Provide the following cluster configuration information and click on **Next** to continue. 

  |**Parameter**| **Description**|
  |-------------|---------------|

>   |**Static Placement** | By default, Palette uses dynamic placement. This creates a new Virtual Private Cloud (VPC) for the cluster that contains two subnets in different Availability Zones (AZs), which is required for EKS cluster deployment. Palette places resources in these clusters, manages the resources, and deletes them when the corresponding cluster is deleted.<br /><br />If you want to place resources into pre-existing VPCs and subnets, enable the **Static Placement** option, and provide the VPCID in the **VPCID** field that displays with this option enabled. You will need to specify two subnets in different Availability Zones (AZs). |

Author: Oops, I forgot to delete that after adding the last sentence. Deleted.

  |**Region** | Use the **drop-down Menu** to choose the AWS region where you would like to provision the cluster.|
  |**SSH Key Pair Name** | Choose the SSH key pair for the region you selected. SSH key pairs must be pre-configured in your AWS environment. This is called an EC2 Key Pair in AWS. The key you select is inserted into the provisioned VMs.|
  |**Cluster Endpoint Access**| This setting provides access to the Kubernetes API endpoint. Select **Private**, **Public** or **Private & Public**. For more information, refer to the [Amazon EKS cluster endpoint access control](https://docs.aws.amazon.com/eks/latest/userguide/cluster-endpoint.html) reference guide.|
  |**Public Access CIDRs** |This setting controls which IP address CIDR ranges can access the cluster. To fully allow unrestricted network access, enter `0.0.0.0/0` in the field. For more information, refer to the [Amazon EKS cluster endpoint access control](https://docs.aws.amazon.com/eks/latest/userguide/cluster-endpoint.html) reference guide.|

>   |**Private Access CIDRs** |This setting controls which private IP address CIDR ranges can access the cluster. Private CIDRs provide a way to specify private, self-hosted, and air-gapped networks or Private Cloud Gateway (PCG) that may be located in other VPCs connected to the VPC hosting the cluster endpoint.<br /><br />To restrict network access, enter the IP address CIDR range that will provide access to the cluster. Although `0.0.0.0/0` is pre-populated in this field, Palette ensures the only IPs that can reach the private endpoint are those within the VPC or any other connected VPCs. For more information, refer to the [Amazon EKS cluster endpoint access control](https://docs.aws.amazon.com/eks/latest/userguide/cluster-endpoint.html) reference guide.|

Reviewer 3: Would it be beneficial to state the difference of using a VPC CIDR vs an open `0.0.0.0/0`. Something along the lines of..

While using `0.0.0.0/0` would allow traffic throughout the VPC and all peered VPCs, specifying the VPC CIDR `10.0.0.0/16` would limit to an individual VPC?

Author: Added it as an example.

>   |**Private Access CIDRs** |This setting controls which private IP address CIDR ranges can access the cluster. Private CIDRs provide a way to specify private, self-hosted, and air-gapped networks or Private Cloud Gateway (PCG) that may be located in other VPCs connected to the VPC hosting the cluster endpoint.<br /><br />To restrict network access, enter the IP address CIDR range that will provide access to the cluster. Although `0.0.0.0/0` is pre-populated in this field, Palette ensures the only IPs that can reach the private endpoint are those within the VPC or any other connected VPCs. For more information, refer to the [Amazon EKS cluster endpoint access control](https://docs.aws.amazon.com/eks/latest/userguide/cluster-endpoint.html) reference guide.|

Reviewer 2: it isn't really palette that ensures only IPs within the VPC or connected VPCs can reach the endpoint. The reason is because the VPC endpoint is not publically routable, so no internet traffic can reach it. Therefore, setting 0.0.0.0/0 here will have the effect of allowing all traffic from within the VPC or any connected VPCs.

Author: Thanks for explaining. I removed "Palette ensures".

kms:DescribeKeys
```
Ensure the IAM role or IAM user can perform the required IAM permissions on the KMS key that will be used for EKS.

You can enable secret encryption during the EKS cluster creation process by toggling the **Enable Encryption** button on and providing the Amazon Resource Name (ARN) of the encryption key. The encryption option is available on the cluster creation wizard's **Cluster Config** page. Review [EKS Cluster Encryption](#eks-cluster-secrets-encryption) for more details.

## Resources

> 

Author: Agree and reduced. I thought we had to add each reference on the page. :)
